astropy/table/index.py

```diff
@@ -0,0 +1,308 @@
+from copy import deepcopy
+import numpy as np
+
+from .bst import BST, RedBlackTree, FastBST, FastRBT
+from .array import SortedArray
+
+'''
+The Index class can use several implementations as its
+engine. Any implementation should implement the following:
+
+__init__([dict] lines) : initializes based on key/row list pairs
+add(key, row) -> None : add (key, row) to existing data
+remove(key, data=None) -> boolean : remove data from self[key], or all of
+                                    self[key] if data is None
+reorder(row) -> None : decrement row numbers after row
+find(key) -> list : list of rows corresponding to key
+range(lower, upper, bounds) -> list : rows in self[k] where k is between
+                               lower and upper (<= or < based on bounds)
+sort() -> list of rows in sorted order (by key)
+replace_rows(row_map) -> None : replace row numbers based on slice
+items() -> list of tuples of the form (key, data)
+
+
+Note: when a Table is initialized from another Table, indices are
+(deep) copied and their columns are set to the columns of the new Table.
+
+Column creation:
+Column(c) -> deep copy of indices
+c[[1, 2]] -> deep copy and reordering of indices
+c[1:2] -> reference
+array.view(Column) -> no indices
+'''
+
+class MaxValue(object):
+    def __gt__(self, other):
+        return True
+
+    def __ge__(self, other):
+        return True
+
+    def __lt__(self, other):
+        return False
+
+    def __le__(self, other):
+        return False
+
+    def __repr__(self):
+        return "MAX"
+
+    __le__ = __lt__
+    __ge__ = __gt__
+    __str__ = __repr__
+
+
+class MinValue(object):
+    def __lt__(self, other):
+        return True
+
+    def __le__(self, other):
+        return True
+
+    def __gt__(self, other):
+        return False
+
+    def __ge__(self, other):
+        return False
+
+    def __repr__(self):
+        return "MIN"
+
+    __le__ = __lt__
+    __ge__ = __gt__
+    __str__ = __repr__
+
+class QueryError(ValueError):
+    pass
+
+class Index:
+    def __init__(self, columns, impl=None, col_dtypes=None, data=None):
+        from .table import Table
+        if data is not None: # create from data
+            self.engine = data.__class__
+            self.data = data
+            self.columns = columns
+            return
+
+        self.engine = impl or FastRBT
+
+        if columns is None: # this creates a special exception for deep copying
+            columns = []
+            lines = []
+        elif len(columns) == 0:
+            raise ValueError("Cannot create index without at least one column")
+        else:
+            num_rows = len(columns[0])
+            table = Table(columns + [[x for x in range(num_rows)]])
+            lines = table.group_by([x.name for x in columns])
```

Comment:
All column attributes should be accessed via the `info` attribute (e.g. `x.info.name`).  This is the mixin-safe idiom.